src/WebJobs.Script.WebHost/Metrics/LinuxContainerLegionMetricsPublisher.cs

```diff
@@ -75,21 +76,7 @@ public LinuxContainerLegionMetricsPublisher(IEnvironment environment, IOptionsMo
             _hostingConfigOptionsOnChangeListener = _hostingConfigOptions.OnChange(OnHostingConfigOptionsChanged);
         }
 
-        public IMetricsLogger MetricsLogger
-        {
-            get
-            {
-                if (_metricsLogger == null)
-                {
-                    if (!Utility.TryGetHostService<IMetricsLogger>(_scriptHostManager, out _metricsLogger))
-                    {
-                        throw new InvalidOperationException($"Unable to resolve {nameof(IMetricsLogger)} service.");
-                    }
-                }
-
-                return _metricsLogger;
-            }
-        }
+        private IMetricsLogger MetricsLogger => _metricsLogger ??= _serviceProvider.GetRequiredService<IMetricsLogger>();
```

Comment:
I think you're changing the scoping here? The `IScriptHostManager` returns services from the current JobHost -- but is this now using the WebHost's services?

Reply:
I believe the webhost scope is the correct scope to get the `IMetricsLogger` instance. `IMetricsLogger` is registered on the webhost level.

https://github.com/Azure/azure-functions-host/blob/701f376f9afbf7177f59791414590a0272934b4f/src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs#L158

Reply:
This service is registered in both layers, and they are the same underlying implementation in the end. The difference with the JobHost copy is that it is a wrapper which suppresses dispose.

https://github.com/Azure/azure-functions-host/blob/deb86165644814c94ca0b568bc698dbcffdbf84e/src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs#L162

If avoiding disposal is not a concern here, then I believe getting it directly from the WebHost layer is fine.

Reply:
I agree that this looks fine. LinuxContainerLegionMetricsPublisher is not attempting to dispose the IMetricsLogger - it's just using the one that it's given.